VEP #169: File-Level Restore (#170)
kubevirt/enhancements: veps/sig-storage/file-restore.md

## Repos
[KubeVirt](https://github.com/kubevirt/kubevirt)

## Design
### Guest-Level Cooperation
The solution is designed to support file-level restore for both Linux and Windows guests. Windows NTFS filesystems require guest cooperation to properly handle filesystem-specific features, including: access control lists (ACL), alternate data streams (ADS), hard links and junction points, file metadata (attributes and timestamps).

Additionally, encrypted volumes (such as BitLocker for Windows) require guest-level access to ensure files are backed up in their encrypted state and properly restored.

### Implementation Approach
* Hotplug volume will be used to attach the VM a backup PVC or a VolumeSnapshot restored PVC
* A helper VirtualMachineGuestCommand CRD will be created by the controller for executing allowed guest commands using the specified transport.
* SSH over VSOCK will be the reference transport, while the API will allow adding support for other transports as well (SSH etc.).

> * The guest OS-specific operations will be warpped in a restricted script or binary. We will provide reference scripts, but it can be implemented by a backup vendor guest agent as well.

Reviewer: "warpped" -> "wrapped"

* As a backup vendor, I would like to offer file level restore capability similar to vSphere

## Repos
[KubeVirt](https://github.com/kubevirt/kubevirt)

## Design
### Guest-Level Cooperation
The solution is designed to support file-level backup and restore for both Linux and Windows guests. Windows NTFS filesystems require guest cooperation to properly handle filesystem-specific features, including: access control lists (ACL), alternate data streams (ADS), hard links and junction points, file metadata (attributes and timestamps).

Additionally, encrypted volumes (such as BitLocker for Windows) require guest-level access to ensure files are backed up in their encrypted state and properly restored.

### Implementation Approach
* Backup and restore are independent operations, and restore can be supported even without the backup support
* Backup and restore are idempotent operations, so the same CRs can be applied whenever one wants to backup or restore
* Hotplug volume will be used to attach the VM a PVC for backup and restore, or VolumeSnapshot restored PVC for restore
* Backup and restore with guest ssh-able target will also be supported

> * SSH over VSOCK will be used to execute guest operations (mount backup volume filesystem, rsync etc.) using a helper  VirtualMachineGuestCommand CRD

Reviewer 2: While VSOCK provides some benefits over the virtual networking, is there a benefit from limiting the design to VSOCK, or would it be possible to have the design more generic, so that TCP could be utilized, too?

> * SSH over VSOCK will be used to execute guest operations (mount backup volume filesystem, rsync etc.) using a helper  VirtualMachineGuestCommand CRD

Author: We use only a minimal set of guest commands (e.g. mkdir, mount, rsync), so I don't think it creates dependency on its internal configuration. As mentioned, since we don't involve a guest agent, the controller creates guest OS specific commands (Linux/Windows).

  * If the PVC does not exist, create it
  * Hotplug the volume
  * Using VirtualMachineGuestCommand
    * Format the volume if needed
    * Mount the backup filesystem
    * Backup the specified files or directories
    * Unmount the backup filesystem
  * Unplug the volume
* If remote target
  * Using VirtualMachineGuestCommand
    * Backup the specified files

### File-Level Restore
When VirtualMachineFileRestore is applied:
* If VolumeSnapShot source, restore it to PVC in the VM namespace
* If PVC source (or VolumeSnapShot restored PVC)
  * Hotplug the volume
  * Using VirtualMachineGuestCommand
    * Mount the backup filesystem
    * Restore the specified files or directories
    * Unmount the backup filesystem
  * Unplug the volume
* If remote source
  * Using VirtualMachineGuestCommand
    * Restore the specified files

## API Examples

For using this feature the HotplugVolumes and VSOCK feature gates must be enabled.

### Backup
#### Directory Backup to PVC
```yaml
apiVersion: filerestore.kubevirt.io/v1alpha1
kind: VirtualMachineFileBackup
metadata:
  name: filebackup1
spec:
  vmiName: fedora
  sourcePath: /home/donald
  target:
    pvc:
      name: filebackup1 # optional

>       size: 1Gi # optional

Author: The original idea was to also support volume expansion of existing backup PVC using the `size` parameter. I don't think we should provision the backup PVC based of what `rsync --dryrun` says, as we want to allow limiting its size. In addition, it may need much more space when the following backup CR is applied. I guess that failing with a clear error specifying the required size is a reasonable approach.

  #targetPath: /home/duck #optional
status:
  phase: InProgress
```
#### Directory Backup to Remote Host
Similar to PVC except the target
```yaml
  target:
    host:
      address: backup-01.us-east.internal # ip or name
```

### Restore
#### Directory Restore from PVC
```yaml
apiVersion: filerestore.kubevirt.io/v1alpha1
kind: VirtualMachineFileRestore
metadata:
  name: filerestore1
spec:
  vmiName: fedora

>   source:

Reviewer 3: is adding an optional backup reference field (pointing to a VirtualMachineFileBackup CR) worth it? i mean If provided, the controller auto-fills the pvc and sourcePath.

    pvc:
      name: filebackup1

>   sourcePath: /home/donald

Author: It is recreated by rsync if possible, and if it fails the `VirtualMachineGuestCommand` will report the failure and the `VirtualMachineFileRestore` will fail as well.

  #targetPath: /home/duck #optional
status:
  phase: Succeeded
```
#### Directory Restore from Volume Snapshot
Similar to PVC except the source
```yaml
  source:
    snapshot:
      name: snap1
```
#### Directory Restore from Remote Host
Similar to PVC except the source
```yaml
  source:
    host:
      address: backup-01.us-east.internal # ip or name
```

### Guest Command

> To perform the actual guest operations the controller will create a VirtualMachineGuestCommand, which will execute the required commands inside the guest via SSH over, for example:

Reviewer 3: i wonder if `VirtualMachineGuestCommand` accepts raw shell commands, there must be an admission controller or Validating Webhook that strictly limits allowed binaries (for exampl only allow rsync, mkdir, mount). Without this, a user could craft a command to execute `dd if=/dev/zero of=/dev/vda`, destroying the VM

## Repos
[KubeVirt](https://github.com/kubevirt/kubevirt)

## Design
### Guest-Level Cooperation
The solution is designed to support file-level restore for both Linux and Windows guests. Windows NTFS filesystems require guest cooperation to properly handle filesystem-specific features, including: access control lists (ACL), alternate data streams (ADS), hard links and junction points, file metadata (attributes and timestamps).

Additionally, encrypted volumes (such as BitLocker for Windows) require guest-level access to ensure files are backed up in their encrypted state and properly restored.

### Implementation Approach
* Hotplug volume will be used to attach the VM a backup PVC or a VolumeSnapshot restored PVC

> * SSH over VSOCK will be used to execute guest operations (mount backup volume filesystem, rsync etc.) using a helper VirtualMachineGuestCommand CRD

Author: Can we `mkdir`/`mount`/`rsync` with `internal-sftp`?

> * SSH over VSOCK will be used to execute guest operations (mount backup volume filesystem, rsync etc.) using a helper VirtualMachineGuestCommand CRD

Reviewer 2: I see, if the PVC required to be mounted directly to the guest VM, mount would be required, but sftp operates at [file level](https://github.com/openssh/openssh-portable/blob/master/sftp.h#L34).

> * SSH over VSOCK will be used to execute guest operations (mount backup volume filesystem, rsync etc.) using a helper VirtualMachineGuestCommand CRD

Reviewer 2: > Is it possible to prevent arbitrary commands being executed inside the VM over the ssh connection? (For example, it may be necessary to install an safe rsync wrapper that restricts the command-line and then only allow that wrapper in .ssh/authorized_keys.)

> Some of these questions depend on the user's security requirements, but often users do not want to give unrestricted ssh access because they consider it a security risk.

Full ack! This is why I would vote for doing all the [permission demanding work in a help pod](https://github.com/kubevirt/enhancements/pull/170#discussion_r2736965476), and restricting the acess to sftp.

> Regarding ssh, does this mean something needs to install the OpenSSH server inside the VM and configure it to listen over vsock?

> The host connects to sshd with which user name (root?) and key? How is the ssh key managed? Is it unique for each VM?

Ack, would be great to have answers to those great quesions in the VEP.

* Since we do not depend on guest agent, our controler creating the VirtualMachineGuestCommand will execute a minimal set of OS-specific (Linux/Windows) commands
* After the specified files are copied, another pass(es) will follow to make sure no files changed during the operation, so we can timestamp it. Unlike fsfreeze this will not guarentee a complete file system consistency.

### File-Level Restore
When VirtualMachineFileRestore is applied:
* If VolumeSnapShot source, restore it to PVC in the VM namespace
* If PVC source (or VolumeSnapShot restored PVC)
  * Hotplug the volume (read-only)
  * Using VirtualMachineGuestCommand
    * Mount the backup filesystem (read-only)
    * Restore the specified files or directories
    * Unmount the backup filesystem
  * Unplug the volume

### File-Level Backup
We also provide a file-level backup functionality. It is useful when the VM user knows what he wants to backup. File-level restore is not dependent on it at all.

When VirtualMachineFileBackup is applied:
* If the target is PVC
  * If the PVC does not exist, create it
  * Hotplug the volume
  * Using VirtualMachineGuestCommand

>     * Format the volume if needed

Reviewer: My understanding is that rsync skips the delta algorithm by default when copying between local files. While it is possible to explicitly enable the delta algorithm on the command-line, that still involves reading the source and destination files and that is expensive for large files. So there is no way to make the backup really efficient with rsync itself.

Efficient solutions are more specialized, like dm-era, btrfs diff, or inotify-based changed block tracking. It makes sense to support a general-purpose solution like rsync or cp, but it may be necessary to support specialized changed block tracking solutions for large file use cases.

>     * Format the volume if needed

Author: Cool. We currently leave the guest-file-restore-tool open, as it will be wrapped in a trusted guest script/binary. rsync was just an initial suggestion.

# VEP #169: File-Level Backup and Restore

## Release Signoff Checklist
Items marked with (R) are required *prior to targeting to a milestone / release*.
- [X] (R) Enhancement issue created, which links to VEP dir in [kubevirt/enhancements] (not the initial VEP PR)
- [ ] (R) Target version is explicitly mentioned and approved
- [ ] (R) Graduation criteria filled

## Overview
Proposal to support virtual machine file-level backup and restore, applying on specific files or directories. This will give users and data protection partners the ability to restore files or subset of user data into a virtual machine.

## Motivation

> Currently, you need to restore an entire VM or a specific volume separately, and retrieve the data from it. This enhancment will allow the user restore the small set of files needed.

Reviewer 4: The motivation for this proposal isn't clear to me.

> Currently, you need to restore an entire VM or a specific volume separately, and retrieve the data from it.

It isn't obvious why this is inherently problematic. Even if we assume this is a limitation, it's unclear why KubeVirt should be responsible for addressing it at the data-plane level.

In all existing tangential areas -- such as VMSnapshot and the upcoming VMBackup feature -- KubeVirt deliberately acts as an orchestrator that relies on third-party implementations; CSI drivers for VMSnapshot, and QEMU/KVM (at least today) for storage-agnostic VMBackup. This separation of concerns has been intentional and effective, and I don’t see a compelling reason to deviate from it here.

Moreover, many backup vendors already provide mature guest-level, file-level backup and file-level restore solutions. If the goal is to enable or improve support for such workflows, it would seem more appropriate for KubeVirt to facilitate their integration rather than reimplementing similar functionality itself.

One possible direction could be a shim or mediation layer. Conceptually it would be similar to something like a CSI in the sense that host-side components (e.g., virt-handler or virt-launcher) would forward user-initiated requests to a guest backup agent.

## Goals
- Provide a declarative API allowing user to partially backup and restore VM data.
- Guest OS agnostic solution, supporting Windows (NTFS) in addition to Linux.
- Backup vendor agnostic solution.

## Non Goals
- Backup file browsing is not supported, but may leverage OADP file serving infrastructure
- Parallel file backups/restores of the same VMs are not supported

## Definition of Users
* VM owners
* Cluster Admins
* Backup vendors

## User Stories
* As a KubeVirt user, I would like to partially restore my VM data with specific files or directories
* As a KubeVirt user, I would like to partially backup my VM data with specific files or directories

> * As a backup vendor, I would like to offer file level restore capability similar to vSphere

Author: We are sharing the VEP with backup vendors to gather their feedback and input.

* As a user of a Linux VM, I would like to backup and restore a single file on a LUKS encrypted volume in my running VM.
* As a user of a Linux VM, I would like to backup and restore multiple files in a single directory on a LUKS encrypted volume in my running VM in a consistent manner.
* As a user of a Windows VM, I would like to backup and restore a single file on a bitlocker encrypted file system in my running VM.
* As a user of a Windows VM, I would like to backup and restore multiple files in a single directory on a bitlocker encrypted file system in my running VM in a consistent manner.

## Repos
[KubeVirt](https://github.com/kubevirt/kubevirt)

## Design
### Guest-Level Cooperation
The solution is designed to support file-level backup and restore for both Linux and Windows guests. Windows NTFS filesystems require guest cooperation to properly handle filesystem-specific features, including: access control lists (ACL), alternate data streams (ADS), hard links and junction points, file metadata (attributes and timestamps).

> Additionally, encrypted volumes (such as BitLocker for Windows) require guest-level access to ensure files are backed up in their encrypted state and properly restored.

Reviewer 5: how are keys managed?

> Additionally, encrypted volumes (such as BitLocker for Windows) require guest-level access to ensure files are backed up in their encrypted state and properly restored.

Reviewer 5: > We put a key file in the VM encrypted volume that we restore to from its encrypted backup PVC/VolumeSnapshot.

So "we" meaning some KubeVirt code somehow gets the encryption key for a disk and store that key in a regular file on the filesystem? We use this key to unlock the restored disk? 

Can you describe this process in detail in the VEP?

> Additionally, encrypted volumes (such as BitLocker for Windows) require guest-level access to ensure files are backed up in their encrypted state and properly restored.

Author: Encrypted volume key management is in the responsibility of the guest file restore helper, looking for them in a specific location. There is no KubeVirt code dealing with the encryption keys, it's stays in the guest. Added this note to the VEP.

### Implementation Approach
* Backup and restore are independent operations, and restore can be supported even without the backup support
* Backup and restore are idempotent operations, so the same CRs can be applied whenever one wants to backup or restore
* Hotplug volume will be used to attach the VM a PVC for backup and restore, or VolumeSnapshot restored PVC for restore

> * Backup and restore with guest ssh-able target will also be supported

Reviewer 6: This assumes that ssh support is enabled in the guest, I assuming this won't work if ssh is not enabled in the guest?

Author: Right, we will fail the operation with an informative error in this case.

* SSH over VSOCK will be used to execute guest operations (mount backup volume filesystem, rsync etc.) using a helper  VirtualMachineGuestCommand CRD
* Since we do not depend  on guest agent, our controler creating the VirtualMachineGuestCommand will execute a minimal set of OS-specific (Linux/Windows) commands
* After the specified files are copied, another pass(es) will follow to make sure no files changed during the operation, so we can timestamp it. Unlike fsfreeze this will not guarentee a complete file system consistency.

### File-Level Backup
When VirtualMachineFileBackup is applied:
* If the target is PVC
  * If the PVC does not exist, create it
  * Hotplug the volume
  * Using VirtualMachineGuestCommand
    * Format the volume if needed
    * Mount the backup filesystem

>     * Backup the specified files or directories

Reviewer 6: How is the backup achieved? something like rsync? Does the guest have to have something installed to do the backup?

    * Unmount the backup filesystem

>   * Unplug the volume

Reviewer 5: I think it would be better to have the backed up files directly on a FileSystem PVC instead of inside a disk image on a PVC. I think we could accomplish that with the following flow:

1. Hotplug [utility volume](https://github.com/kubevirt/enhancements/blob/main/veps/sig-storage/utility-volumes.md) in virt launcher
2. Run rsync client in virt-handler. Connect to guest over ssh + vsock. Write to utility volume

I'm not exactly sure how to configure rsync to use ssh+vsock but it seems to be possible by using `ProxyCommand` with ssh and netcat (or some other proxy)

> * If remote target

Reviewer 5: Guest would need ssh key copied over to do remote backup

  * Using VirtualMachineGuestCommand
    * Backup the specified files

### File-Level Restore
When VirtualMachineFileRestore is applied:
* If VolumeSnapShot source, restore it to PVC in the VM namespace
* If PVC source (or VolumeSnapShot restored PVC)
  * Hotplug the volume (read-only)
  * Using VirtualMachineGuestCommand
    * Mount the backup filesystem (reda-only)
    * Restore the specified files or directories
    * Unmount the backup filesystem
  * Unplug the volume
* If remote source
  * Using VirtualMachineGuestCommand
    * Restore the specified files

## API Examples

> For using this feature the DeclarativeHotplugVolumes and VSOCK feature gates must be enabled.

Reviewer 6: ```suggestion
In order to use this feature both the DeclarativeHotplugVolumes and VSOCK feature gates must be enabled.
```

For access control, VirtualMachineFileBackup/Restore will have similar RBAC as VirtualMachineSnapshot/Restore, so the same service accounts are authorized to create them.

### Backup
#### Directory Backup to PVC
```yaml
apiVersion: filerestore.kubevirt.io/v1alpha1
kind: VirtualMachineFileBackup
metadata:

>   name: filebackup1

Author: Sure, namespace scoped. Fixing.

spec:
  vmiName: fedora

>   sourcePath: /home/donald # optional, otherwise manual copy

Author: Following the same flow until backup filesystem is mounted, and letting the user manually do the actual file copy, and when he's done with it he deletes the CR, which does the umount and volume unplug.

  target:
    pvc:
      name: filebackup1    # optional, otherwise use CR name
      size: 1Gi            # optional

>   #targetPath: /home/duck  # optional

Reviewer 6: I am assuming if omitted, that the targetPath will match the source path but from the base of the volume?

>   #targetPartition: 1      # optional

Reviewer 6: I don't understand this, are we planning on more than one partition per PVC?

status:

>   phase: InProgress

Author: Sure, will add these and think of other relevant ones.

```
#### Directory Backup to Remote Host
Similar to PVC except the target
```yaml
  target:
    host:
      address: backup-01.us-east.internal # ip or name
```

### Restore
#### Directory Restore from PVC
```yaml
apiVersion: filerestore.kubevirt.io/v1alpha1
kind: VirtualMachineFileRestore
metadata:
  name: filerestore1
spec:
  vmiName: fedora
  source:
    pvc:
      name: filebackup1
  sourcePath: /home/donald # optional, otherwise manual copy
  #sourcePartition: 1      # optional
  #targetPath: /home/duck  # optional
status:
  phase: Succeeded
```

> #### Directory Restore from Volume Snapshot

Reviewer 6: I don't see an explanation of how we arrived at this snapshot. Was it an entire VMSnapshot or something else? How was this snapshot created?

Similar to PVC except the source
```yaml
  source:
    snapshot:
      name: snap1
```
#### Directory Restore from Remote Host
Similar to PVC except the source
```yaml
  source:
    host:
      address: backup-01.us-east.internal # ip or name
```

### Guest Command
To perform the actual guest operations the controller creates a VirtualMachineGuestCommand, which executes the required commands inside the guest via SSH over VSOCK. We currently do not want to support running arbitrary commands on the guest, so only the `kubevirt-controller` service account has RBAC for creating VirtualMachineGuestCommand.

```yaml
apiVersion: guestcommand.kubevirt.io/v1alpha1

> kind: VirtualMachineGuestCommand

Author: See [previous](https://github.com/kubevirt/enhancements/pull/170#discussion_r2702221127) answer.

metadata:
  name: cmd1
spec:
  vmiName: fedora
  transport:
    vsock:
      name: fedora

>   command:

Reviewer 6: I am super uncomfortable with being able to execute arbitrary commands from a k8s resource, even if the RBAC provides some protection. Would it make more sense to make a list of possible acceptable commands, and the controller takes the name of the command and arguments to that command (like the paths and names of the files) and executes only the acceptable commands. We would need to make sure we sanitize the arguments as well.

  - /bin/sh
  - -c

>   - rsync -avR /home/donald /backup/

Reviewer 6: This assumes that rsync is installed in the guest OS if I understand correctly. This might not always be the case especially with windows guests.

Author: Answered [here](https://github.com/kubevirt/enhancements/pull/170#issuecomment-3765320203).

status:
  phase: Succeeded

>   reason: SuccessfulExecution

Reviewer 6: Some conditions here would be good as well. Might not be able to execute the command due to missing binaries etc. Some conditions to indicate issues and help trouble shoot will be helpful.

Author: Agree, will add some relevant ones.

spec:
  user: root
  sshKeySecret:
    key: ssh-privatekey
    name: my-ssh-key-secret
  port: 2222
  useTLS: false
```

## SSH over VSOCK
Guest SSHD needs to be configured to listen for incoming connections on a VSOCK. In Linux this can be done by `systemd-ssh-generator`, `socat` etc. Windows supports VSOCK via `virtio-win`, so we need to install viosock driver, viosock-tcp bridge service, OpenSSH server and Cloudbase-Init. `virt-handler` connects the guest `sshd` with restore-user, which is either a user with specific `sudo` permissions for file operations, or `root`. We need to decide if this restore-user is common or unique for each VM. The user and public key are created in the guest via VM spec cloudInit userData. We do not want to give unrestricted ssh access, so the private key is stored in a k8s Secret that only the `kubevirt-handler` service account has RBAC to access. To prevent arbitrary commands being executed inside the VM over the ssh connection we will only support execution of a specifically-named safe wrapper installed on the guest, doing the actual restore (via `rsync` etc), restricting the command-line, and allow only that wrapper in .ssh/authorized_keys.

> ## Alternatives

Author: How would you open the encrypted volume in the additional pod before you mount its file system?

# VEP #169: File-Level Backup and Restore

## Release Signoff Checklist
Items marked with (R) are required *prior to targeting to a milestone / release*.
- [X] (R) Enhancement issue created, which links to VEP dir in [kubevirt/enhancements] (not the initial VEP PR)
- [ ] (R) Target version is explicitly mentioned and approved
- [ ] (R) Graduation criteria filled

## Overview
Proposal to support virtual machine file-level backup and restore, applying on specific files or directories. This will give users and data protection partners the ability to restore files or subset of user data into a virtual machine.

## Motivation
Currently, you need to restore an entire VM or a specific volume separately, and retrieve the data from it. This enhancment will allow the user restore the small set of files needed.

## Goals
- Provide a declarative API allowing user to partially backup and restore VM data.

> - Guest OS agnostic solution, supporting Windows (NTFS) in addition to Linux.

Reviewer 5: Nos as concerned about ssh running or rsync existing as I am about ssh listening on VSOCK

- Backup vendor agnostic solution.

## Non Goals
- Backup file browsing is not supported, but may leverage OADP file serving infrastructure
- Parallel file backups/restores of the same VMs are not supported

## Definition of Users
* VM owners
* Cluster Admins
* Backup vendors

## User Stories
* As a KubeVirt user, I would like to partially restore my VM data with specific files or directories
* As a KubeVirt user, I would like to partially backup my VM data with specific files or directories
* As a backup vendor, I would like to offer file level restore capability similar to vSphere
* As a user of a Linux VM, I would like to backup and restore a single file on a LUKS encrypted volume in my running VM.
* As a user of a Linux VM, I would like to backup and restore multiple files in a single directory on a LUKS encrypted volume in my running VM in a consistent manner.
* As a user of a Windows VM, I would like to backup and restore a single file on a bitlocker encrypted file system in my running VM.
* As a user of a Windows VM, I would like to backup and restore multiple files in a single directory on a bitlocker encrypted file system in my running VM in a consistent manner.

## Repos

> [KubeVirt](https://github.com/kubevirt/kubevirt)

Reviewer 7: I agree with @Acedus on this one
> Even if we assume this is a limitation, it's unclear why KubeVirt should be responsible for addressing it at the data-plane level.

I think we have to ask ourselves where we draw the line wrt "core" kubevirt features;
As we see in this VEP, there's already some concern with being in this problem domain to begin with
(guest binaries etc) so to me that already implies that we should at least consider separate project

spec:
  vmiName: fedora
  source:
    snapshot:
      name: snap1
  sourcePath: /home/donald
```

#### Manual Restore
When `sourcePath` is not provided, we allow manual restore from the hotplugged backup volume mounted filesystem using any preferred guest tool. When done, the user deletes the CR, so we unmount the filesystem and unplug the volume.
```yaml
apiVersion: filerestore.kubevirt.io/v1alpha1
kind: VirtualMachineFileRestore
metadata:
  name: filerestore1
  namespace: ns1
spec:
  vmiName: fedora
  source:
    snapshot:
      name: snap1
```

### Guest Command
To perform the actual guest operations the controller creates a VirtualMachineGuestCommand, which executes the required commands inside the guest via the specified transport (e.g. SSH over VSOCK). We currently do not want to support running arbitrary commands on the guest, so only the `kubevirt-controller` service account has RBAC for creating VirtualMachineGuestCommand. We will define an allow-list of possible acceptable commands. The controller will take the name of the command and its arguments (e.g. paths and names of files), and execute only the acceptable commands. Arguments will be sanitized as well.

> 

Reviewer 5: I'm with @vladikr in that a declarative API is not ideal executing commands. 

Not using the subresource API would require setting up a lot of plumbing to handle auth and routing to the appropriate virt-handler though. 

Would love to avoid this all together. I still don't fully understand the requirement that these commands execute over vsock+ssh. To support VMs without networking? How common is that? Why not ssh over the network? At least initially and see if there is demand for vsock?

> 

Author: (Back from holiday) I agree with your feedback, and will update the VEP accordingly, dropping the guest command CRD and using an internal RPC instead. The ssh reference will be generalized, mentioning we can start initially with ssh over the network, and add vsock support if there is demand.

```yaml
apiVersion: guestcommand.kubevirt.io/v1alpha1
kind: VirtualMachineGuestCommand
metadata:
  name: cmd1
  namespace: ns1
spec:
  vmiName: fedora
  config: myconfig
  command:
  - filerestore --serial "fedora-filebackup1-backup" --mount-path "/backup" --source-path "/home/donald"
status:
  conditions:
  - type: Completed
    status: "False"
    reason: MissingBinary
    message: "Required binary 'filerestore' not found in guest."
    lastTransitionTime: "2026-01-01T01:23:45Z"
  - type: Failed
    status: "True"
    reason: MissingBinary
    message: "Required binary 'filerestore' not found in guest."
    lastTransitionTime: "2026-01-01T01:23:45Z"
---
apiVersion: guestcommand.kubevirt.io/v1alpha1
kind: VirtualMachineGuestCommandConfig
metadata:
  name: myconfig
  namespace: ns1
config
  transport: vsock
  user: filerestore
  port: 22
  sshKeySecret:
    secretName: my-secret # alternative: `qemuGuestAgent: {}` for temporary secret key injected via qemu-guest-agent
---
kind: Secret
apiVersion: v1
metadata:
  name: my-secret
  namespace: ns1
data:
  key: c3...
type: Opaque
```

## SSH over VSOCK

> Guest SSHD needs to be configured to listen for incoming connections on a VSOCK. In Linux this can be done by `systemd-ssh-generator`, `socat` etc. Windows supports VSOCK via `virtio-win`, so we need to install viosock driver, viosock-tcp bridge service and OpenSSH server. `virt-handler` connects the guest `sshd` with restore-user, which is either a user with specific `sudo` permissions for file operations, or `root`. We need to decide if this restore-user is common or unique for each VM. We do not want to give unrestricted ssh access, so the private key is stored in a k8s Secret that only the `kubevirt-handler` service account has RBAC to access. To prevent arbitrary commands being executed inside the VM over the ssh connection we will only support execution of a specifically-named trusted file restore helper in the guest, doing the actual restore (`mount`, `rsync` etc), restricting the command-line, and allow only that helper in `.ssh/authorized_keys`.

Reviewer 4: > We do not want to give unrestricted ssh access, so the private key is stored in a k8s Secret that only the `kubevirt-handler` service account has RBAC to access.

I'm a bit confused, is this secret the one specified by the user as part of `VirtualMachineGuestCommandConfig`?

Reviewer 4: AFAIK tenant secret read permissions are a no go, how do we plan on handling that?

## Release Signoff Checklist
Items marked with (R) are required *prior to targeting to a milestone / release*.
- [X] (R) Enhancement issue created, which links to VEP dir in [kubevirt/enhancements] (not the initial VEP PR)
- [ ] (R) Target version is explicitly mentioned and approved
- [ ] (R) Graduation criteria filled

## Overview
Proposal to support virtual machine file-level restore, applying to specific files or directories. This will give users and data protection partners the ability to restore files or a subset of user data into a virtual machine.

## Motivation
Currently, you need to restore an entire VM or a specific volume separately, and retrieve the data from it. This enhancement will allow the user to restore the small set of files needed.

## Goals
- Provide a declarative API allowing users to partially restore VM data.
- A guest-OS-agnostic solution, supporting Windows (NTFS) in addition to Linux.
- A backup-vendor-agnostic solution.

## Non Goals
- Creating backup PVCs or volume snapshots, indexing their content, browsing them offline to locate specific items, and making them accessible in the appropriate namespace, are all the responsibility of the backup vendor.

> - Although reference guest-side file restore helpers will be provided, backup vendors are encouraged to provide their own customized and optimized helpers.

Reviewer 3: > backup vendors are encouraged to provide their own customized and optimized helpers.

this is a bit confusing to me considering this is under "Non Goals" ... we do or do not encourage vendors to provide customized helpers?

if we do, how would we know how to run it? should this be (somehow) defined in the `VirtualMachineFileRestore` CR? is there a clear definition of how such customized helper communicates?

Author: Rephrased this non-goal line to make it clearer, and also mentioned it in the "Guest command execution" section.

### Implementation Approach
* We introduce the `VirtualMachineFileRestore` CRD, which is namespace-scoped and created in the VirtualMachine namespace. Access control is enforced through the relevant RBAC rules.
* A hotplug volume will be used to attach a backup PVC or a VolumeSnapshot-restored PVC to the VM. The `DeclarativeHotplugVolumes` feature gate must be enabled.
* SSH over the network will be used for specific command execution on the guest.
* The guest OS-specific operations will be wrapped in a restricted file-restore guest helper script or binary. We will provide reference scripts, but the helper can be implemented by a backup vendor as well.
* Encrypted volume key management is in the responsibility of the guest file restore helper, looking for them in a specific location (e.g. next to the helper, or in the VM encrypted volume that we restore to from its encrypted backup PVC/VolumeSnapshot).

### File-Level Restore
virt-controller:
* Watch `VirtualMachineFileRestore`
* If the source is a VolumeSnapshot, restore it to a PVC in the VM namespace
* If the source is a PVC in another namespace, clone it to a PVC in the VM namespace
* Hotplug the volume (read-only)

> * Label `VirtualMachineFileRestore` with the target node (e.g. `kubevirt.io/target-node=node01`).

Reviewer 4: Why is this needed?

Author: Only for making virt-handler life easier using filtered informer to process only restores to local VMIs. Can we do it cleaner?

* When the restore completes or the CR is deleted:
  * Unplug the volume

>   * Delete the PVC if temporarily created

Reviewer 4: In which scenario would a PVC be temporarily created?

Author: In restoring from a snapshot or from a PVC in another namespace.

virt-handler:
* Watch `VirtualMachineFileRestore` using a filtered informer to process only restores to local VMIs.

> * SSH into the guest and execute the restore helper with the relevant arguments.

Reviewer 4: Is there a reason for executing the file restore helper inside the guest? For example, can't encrypted volumes be unlocked via a key provided in the form of a secret for example? Alternatively, can we prepare a temporary volume that only contains the directory structure with the relevant files to restore via a separate utility pod with the sole purpose of doing that, and once that prep work is done hotplug and copy whatever's required into the guest?

Author: As already discussed here, we do not want to involve kubevirt in guest OS specifics, such as unlocking a Windows Bitlocked backup volume, or NTFS specifics as mentioned in the Guest-Level Cooperation section. Creating a temporary recovery-dedicated volume is sure a valid application of the solution, but not for the Windows Bitlocked NTFS backup volume.

* Update status upon completion (whether succeeded or failed) or timeout.

Guest file restore helper:
* Unlock the volume if encrypted (key file in target volume)
* Mount the backup filesystem (read-only)
* If automated restore:
  * Restore the specified files or directories
  * Unmount the backup filesystem

## API Examples

### Directory Restore from PVC
```yaml
apiVersion: filerestore.kubevirt.io/v1alpha1
kind: VirtualMachineFileRestore
metadata:
  name: filerestore1
  namespace: ns1
spec:
  target:
    apiGroup: kubevirt.io
    kind: VirtualMachine
    name: "fedora"
  source:
    pvc:
      name: filebackup1
      #namespace: ns2

>   sourcePath: /home/donald # optional

Reviewer 3: just making sure i understand, sourcePath is on top of the defined --mount-path?

for example if --mount-path of the hotplugged volume is /mnt/vol, then the full path inside the guest will be /mnt/vol/home/donald?

Author: Right, and it will be restored to `/home/donald`.

```yaml
apiVersion: filerestore.kubevirt.io/v1alpha1
kind: VirtualMachineFileRestore
metadata:
  name: filerestore1
  namespace: ns1
spec:
  target:
    apiGroup: kubevirt.io
    kind: VirtualMachine
    name: "fedora"
  source:
    snapshot:
      name: snap1
```

### Direct restore from external storage (TBD)
For some backup vendors, backup data is not stored on the same cluster and is often moved to external storage like S3, requiring a two-step copy process for restoration, as we expect the backup data on an accessible PVC or volume snapshot within the cluster. To solve the double-copy process, we should support direct file transfer:
```yaml
apiVersion: filerestore.kubevirt.io/v1alpha1
kind: VirtualMachineFileRestore
metadata:
  name: filerestore1
  namespace: ns1
spec:
  target:
    apiGroup: kubevirt.io
    kind: VirtualMachine
    name: "fedora"
  source:
    remote:
      name: s3_backup
      bucket: buck1
  sourcePath: /home/donald # optional
  #targetPath: /home/duck  # optional
```
This reference example assumes the guest helper uses `rclone` and the remote was already configured via `rclone config`.

## Guest command execution

> `virt-handler` connects to the guest `sshd` as restore-user, a user restricted to `sudo` execution of only the trusted file-restore helper (via `.ssh/authorized_keys` and `/etc/sudoers.d`). `virt-operator` generates an SSH key pair and stores it in a Secret. The public key can be added to the guest either manually or propagated by `virt-controller` patching the VM `accessCredentials.sshPublicKey` with `propagationMethod: qemuGuestAgent` so it is injected via qemu-guest-agent. `virt-handler` mounts the private key Secret at `/etc/kubevirt/ssh/id_rsa` and uses it for file-restore SSH connections. If we find that private key compromise is an issue, we can create an ephemeral key pair per SSH connection.

Reviewer 4: > If we find that private key compromise is an issue, we can create an ephemeral key pair per SSH connection.

Consider using SSH certificates instead? https://man.openbsd.org/ssh-keygen.1#CERTIFICATES

### Guest file restore helper CLI

> For covering the mentioned file restore flows, a guest helper CLI should support:

Reviewer 3: as stated in line 17
> A guest-OS-agnostic solution, supporting Windows (NTFS) in addition to Linux.

and line 22
> reference guest-side file restore helpers will be provided 

filerestore.sh is a "simple" helper we provide? how?

to be OS-agnostic, how would such given helper will look like for windows? exe file? i think more info on the helper for windows would be helpful.

Author: We already presented a POC with both Linux and Windows guest helpers. For Windows we used a Powershell script similar to the Linux bash script with few adjustments. Both scripts align to the CLI spec and implement the logic mentioned, and therefore the kubevirt side is kept guest-OS-agnostic.

* Restore (automatic):\
  `filerestore.sh restore --serial <SERIAL> --mount-path <PATH> --source-path <PATH>`\
* Restore (manual - mount only):\
  `filerestore.sh restore --serial <SERIAL> --mount-path <PATH>`
* Cleanup (unmount and remove mount point):\
  `filerestore.sh cleanup --mount-path <PATH>`

> `SERIAL` is the string that identifies the hotplugged disk inside the guest.

Reviewer 4: I assume we'll be providing that serial?

Author: Yep,  the controller sets the Disk serial when it hotplugs the volume.
